kore/src/Kore/Step/Simplification/OverloadSimplifier.hs

```diff
@@ -103,16 +137,98 @@ mkOverloadSimplifier overloadGraph InjSimplifier {isOrderedInj, injectTermTo} =
     unifyOverloadWithinBound injProto s1 s2 topSort =
         headMay withinBound
       where
-        injProtoTop = injProto { injTo = topSort }
         withinBound =
-            map (fmap
-                    (\inj ->
-                        if injFrom inj == injTo inj then Nothing else Just inj
-                    )
-                )
-            . filter (isOrderedInj . snd)
-            . map (\x -> (x, injProtoTop { injFrom = resultSort x }))
+            filterOverloads injProto topSort
             $ OverloadGraph.commonOverloads overloadGraph s1 s2
 
+    unifyOverloadWithSortWithinBound
+        :: Symbol
+        -> Inj ()
+        -> Either String (Maybe (Pair (Symbol, Maybe (Inj ()))))
+    unifyOverloadWithSortWithinBound sym injProto@Inj { injFrom, injTo } =
+        case overloads of
+            Left err -> Left err
+            Right l ->
+                let l' =
+                        [ Pair (s, inj) (s', inj')
+                        | (s, inj, Just (s', inj')) <- l
+                        ]
+                    third (Pair _ (s, _)) = s
+                    mm' = findMaxOverload third l'
+                in case mm' of
+                    Nothing -> Right Nothing
+                    Just m' ->
+                        if checkMaxOverload third m' l'
+                            then Right (Just m')
+                            else errorAmbiguousOverloads (map third l')
+
+      where
+        overloads =
+            traverse ( \(s,inj) ->
+                getOverloadedWithinSort injProto s injFrom >>=
+                  \r -> return (s, inj, r)
+                )
+                overloadings
+        overloadings =
+            filterOverloads injProto injTo . Set.toList
+            $ OverloadGraph.getOverloading overloadGraph sym
+
+    getOverloadedWithinSort
+        :: Inj ()
+        -> Symbol
+        -> Sort
+        -> Either String (Maybe (Symbol, Maybe (Inj ())))
+    getOverloadedWithinSort injProto sym topSort
+        | null overloads = Right Nothing
+        | Just m <- findMaxOverload fst overloads
+        , checkMaxOverload fst m overloads
+        = Right (Just m)
+        | otherwise = errorAmbiguousOverloads (map fst overloads)
+      where
+        overloads =
+            filterOverloads injProto topSort . Set.toList
+            $ OverloadGraph.getOverloaded overloadGraph sym
+
+    updateMaxOverload :: (a -> Symbol) -> a -> Maybe a -> Maybe a
+    updateMaxOverload _ x Nothing = Just x
+    updateMaxOverload f x mm@(Just m)
+        | sm `isOverloadingOrEqual` sx = mm
+        | sx `isOverloading` sm = Just x
```

Comment:
We should introduce a `compareOverloading :: Symbol -> Symbol -> Maybe Ordering` function to streamline this logic.